panoptes_client/caesar.py

```diff
@@ -0,0 +1,126 @@
+from panoptes_client.panoptes import Panoptes
+
+
+class Caesar(object):
+    EXTRACTOR_TYPES = ['blank', 'external', 'question', 'survey', 'who', 'pluck_field', 'shape']
+    REDUCER_TYPES = ['consensus', 'count', 'placeholder', 'external', 'first_extract', 'stats', 'unique_count', 'rectangle', 'sqs']
+    RULE_TO_ACTION_TYPES = {
+        'subject': ['retire_subject', 'add_subject_to_set', 'add_to_collection', 'external'],
+        'user': ['promote_user']
+    }
+
+    def __init__(self, endpoint='https://caesar-staging.zooniverse.org', redirect_url='https://caesar.zooniverse.org/auth/zooniverse/callback'):
+        self.endpoint = endpoint
+        self.headers = {
+            'Accept': 'application/json'
+        }
+
+    def http_get(self, *args, **kwargs):
+        kwargs['endpoint'] = self.endpoint
+        kwargs['headers'] = self.headers
+        return Panoptes.client().get(*args, **kwargs)
+
+    def http_post(self, *args, **kwargs):
+        kwargs['endpoint'] = self.endpoint
+        kwargs['headers'] = {
+            'Accept': 'application/json',
+            'Content-Type': 'application/json'
+        }
+        return Panoptes.client().post(*args, **kwargs)
+
+    def http_put(self, *args, **kwargs):
+        kwargs['endpoint'] = self.endpoint
+        return Panoptes.client().put(*args, **kwargs)
+
+    def http_delete(self, *args, **kwargs):
+        kwargs['endpoint'] = self.endpoint
+        return Panoptes.client().delete(*args, **kwargs)
+
+    def get_reductions_by_workflow_and_subject(self, workflow_id, subject_id):
+        return self.http_get(f'workflows/{workflow_id}/subjects/{subject_id}/reductions')
+
+    def get_workflow_extractors(self, workflow_id):
+        return self.http_get(f'workflows/{workflow_id}/extractors')
+
+    def get_workflow_reducers(self, workflow_id):
+        return self.http_get(f'workflows/{workflow_id}/reducers')
+
+    def get_extracts_by_workflow_and_subject(self, workflow_id, subject_id):
+        return self.http_get(f'workflows/{workflow_id}/extractors/extractor/extracts', params={'subject_id': subject_id})
+
+    def get_extractor_by_workflow_and_id(self, workflow_id, extractor_id):
+        return self.http_get(f'workflows/{workflow_id}/extractors/{extractor_id}')
+
+    def add_workflow(self, workflow_id):
+        return self.http_post('workflows', json={'workflow': {'id': workflow_id}})
+
+    def create_workflow_extractor(self, workflow_id, extractor_key, extractor_type, task_key='T0', other_extractor_attributes=None):
+        self.validate_extractor_type(extractor_type)
+        if other_extractor_attributes is None:
+            other_extractor_attributes = {}
```

Comment:
The default for `other_extractor_attributes` could just be `{}` instead of `None` I think?

Reply:
I think so too. I originally had it his way but was sniffed out by hound linter. I can bring it back

Reply:
Ah ok, so Hound is right it seems: https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python If you use a dict, it creates one dict at runtime and reuses it for each invocation of the function.